worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,247 @@
+func (s *EventQueue) Kill() {
+	s.tomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (s *EventQueue) Wait() error {
+	return s.tomb.Wait()
+}
+
+func (s *EventQueue) unsubscribe(subscriptionID int) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	sub, found := s.subscriptions[subscriptionID]
+	if !found {
+		return
+	}
+
+	for topic := range sub.topics {
+		var updatedFilters []*eventFilter
+		for _, filter := range s.subscriptionsByNS[topic] {
+			if filter.subscriptionID == subscriptionID {
+				continue
+			}
+			updatedFilters = append(updatedFilters, filter)
+		}
+		s.subscriptionsByNS[topic] = updatedFilters
+	}
+
+	delete(s.subscriptions, subscriptionID)
+
+	// Close the subscription channel after a dispatch. If a unsubscription
+	// happens during a dispatch, then it clauses the event queue to potentially
+	// panic. Removal of the subscription from the queue is synchronous, so
+	// the subscription can no longer be used when another dispatch occurs. The
+	// closing of the channel then becomes asynchronous and can be closed during
+	// the next selection.
+	s.tomb.Go(func() error {
+		// Ensure that we give up if the tomb is dying.
+		select {
+		case <-s.tomb.Dying():
+			return tomb.ErrDying
+		case s.actions <- func() {
+			close(sub.changes)
+		}:
+		}
+		return nil
+	})
+}
+
+func (s *EventQueue) loop() error {
+	defer func() {
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+
+		for _, sub := range s.subscriptions {
+			close(sub.changes)
+		}
+		s.subscriptions = make(map[int]*subscription)
+	}()
+
+	for {
+		var (
+			ch changestream.ChangeEvent
+			ok bool
+		)
+		select {
+		case <-s.tomb.Dying():
+			return tomb.ErrDying
+		case action := <-s.actions:
+			action()
+			continue
+		case ch, ok = <-s.stream.Changes():
+			// If the stream is closed, we expect that a new worker will come
+			// again using the change stream worker infrastructure. In this case
+			// just ignore and close out.
+			if !ok {
+				s.logger.Infof("change stream change channel is closed")
+				return nil
+			}
+		}
+
+		// The dispatch is done in two stages to avoid holding the lock for the
+		// duration of the dispatch.
+		subs := s.gatherSubscriptions(ch)
+		if err := s.dispatchChange(subs, ch); err != nil {
+			return errors.Trace(err)
+		}
+	}
+}
+
+func (s *EventQueue) gatherSubscriptions(ch changestream.ChangeEvent) []*subscription {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	var subs []*subscription
+	for _, subOpt := range s.subscriptionsByNS[ch.Namespace()] {
+		if (ch.Type() & subOpt.changeMask) == 0 {
+			if s.logger.IsTraceEnabled() {
+				s.logger.Tracef("ignoring change: %v", ch)
+			}
+			continue
+		}
+
+		if !subOpt.filter(ch) {
+			if s.logger.IsTraceEnabled() {
+				s.logger.Tracef("filtering out change: %v", ch)
+			}
+			continue
+		}
+
+		if s.logger.IsTraceEnabled() {
+			s.logger.Tracef("dispatching change: %v", ch)
+		}
+
+		subs = append(subs, s.subscriptions[subOpt.subscriptionID])
```

Comment:
I'd almost suggest we randomise this subs slice, as it will weed out ordering issues. Although the queue is blocking,  each subscriber blocks only until is sees the published change event not until it is "finished" with the change. We don't want to guarantee ordering or serialisation of event side-effects.